web/pdf_history.js

```diff
@@ -13,21 +13,455 @@
+   */
+  forward() {
+    if (!this.initialized || this._popStateInProgress) {
+      return;
+    }
+    let state = window.history.state;
+    if (this._isValidState(state) && state.uid < (this._uid - 1)) {
+      window.history.forward();
+    }
+  }
+
+  /**
+   * @returns {boolean} Indicating if the user is currently moving through the
+   *   browser history, useful e.g. for skipping the next 'hashchange' event.
+   */
+  get popStateInProgress() {
+    return this.initialized &&
+           (this._popStateInProgress || this._blockHashChange > 0);
+  }
+
+  /**
+   * @private
+   */
+  _pushOrReplaceState(destination, forceReplace = false) {
+    let shouldReplace = forceReplace || !this._destination;
+    let newState = {
+      fingerprint: this.fingerprint,
+      uid: shouldReplace ? this._currentUid : this._uid,
+      destination,
+    };
+
+    if (typeof PDFJSDev !== 'undefined' && PDFJSDev.test('CHROME') &&
+        window.history.state && window.history.state.chromecomState) {
+      // history.state.chromecomState is managed by chromecom.js.
+      newState.chromecomState = window.history.state.chromecomState;
+    }
+    this._resetSomeInternalState(destination, newState.uid);
+
+    if (shouldReplace) {
+      if (typeof PDFJSDev !== 'undefined' &&
+          PDFJSDev.test('FIREFOX || MOZCENTRAL')) {
+        // Providing the third argument causes a SecurityError for file:// URLs.
+        window.history.replaceState(newState, '');
+      } else {
+        window.history.replaceState(newState, '', document.URL);
+      }
+    } else {
+      if (typeof PDFJSDev !== 'undefined' &&
+          PDFJSDev.test('FIREFOX || MOZCENTRAL')) {
+        // Providing the third argument causes a SecurityError for file:// URLs.
+        window.history.pushState(newState, '');
+      } else {
+        window.history.pushState(newState, '', document.URL);
+      }
+    }
+
+    if (typeof PDFJSDev !== 'undefined' && PDFJSDev.test('CHROME') &&
+        top === window) {
+      // eslint-disable-next-line no-undef
+      chrome.runtime.sendMessage('showPageAction');
+    }
+  }
+
+  /**
+   * @private
+   */
+  _tryPushCurrentPosition() {
+    if (!this._position) {
+      return;
+    }
+    let position = this._position;
+
+    if (!this._destination) {
+      this._pushOrReplaceState(position);
+      return;
+    }
+    if (this._destination.hash === position.hash) {
+      return; // The current document position has not changed.
+    }
+    if (!this._destination.page) {
+      // `this._destination` was set through the user changing the hash of
+      // the document. Do not add `this._position` to the browser history,
+      // to avoid "flooding" it with lots of (nearly) identical entries,
+      // since we cannot ensure that the document position has changed.
+      return;
+    }
+
+    let forceReplace = false;
+    if (this._destination.page === position.first ||
```

Comment:
What is this `first` entry exactly? Is there a more descriptive name for it, or could you comment it?

Reply:
It's the *first* visible page (in the viewer), as provided by the `updateviewarea` event. Please refer to https://github.com/mozilla/pdf.js/pull/8775/commits/1f41a595995baedfd9265ad93e226ba0a5a89c0c#diff-3b4ebbe223f93c9123a754b4845844e7R347.
 
If you look a the code in `PDFViewer`, see https://github.com/mozilla/pdf.js/blob/f54dfc63dcb116c1f97577a786372bf4c2fd0ef3/web/pdf_viewer.js#L704 you'll notice that the `pageNumber` parameter there corresponds to the *first* page.

Since, in the viewer, the first visible page often differs from the *most* visible page (i.e. the one shown in the pageNumber input). Compare what's displayed in the pageNumber input with the page number used by the viewBookmark button as you scroll around in the viewer.

I want to prevent issues with multiple (almost) identical history entries being added to the browser history, by only updating it when it's definitely clear that page we're now on is *not* the same as the one currently in the browser history.

Does this help? Based on the above, how do we make this situation clearer in code?